File: leetcode/1680.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        res = 0
        MOD = 1000000007
        for i in range(1, n+1):
            # print(self.calNumBit(i))
            length = self.calNumBit(i)
            res = (res << length) + i
        return res % MOD
    
    def calNumBit(self, n):
        i = 0
        while n > 0:
            n //=2
            i+=1
        return i


if __name__ == "__main__":
    so = Solution()
    n = 100000
    res = so.concatenatedBinary(n)
    print(res)

File: leetcode/1680.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int concatenatedBinary(int n) { 
        long long ans = 0;
        int MOD = 1e9 + 7;
        for (int i = 1; i <= n; i++) {
            int len = calNumBit(i);
            ans = ((ans << len) % MOD) + i;
        }
        return ans;

    }
    
    int calNumBit(int n) {
        int i = 0;
        while (n > 0) {
            n /= 2;
            i++;
        }
        return i;
    }
};

int main() {
    int n = 12;
    Solution so;
    int res = so.concatenatedBinary(n);
    cout << res;
}